pkg/proxyserver/cache/clusterview.go

```diff
@@ -0,0 +1,342 @@
+	return l.versioner.LastSyncResourceVersion()
+}
+
+type AuthCache struct {
+	allResources             sets.String
+	allUsers                 sets.String
+	allGroups                sets.String
+	clusterRoleLister        SyncedClusterRoleLister
+	clusterRolebindingLister SyncedClusterRoleBindingLister
+
+	lastSyncResourceVersioner       LastSyncResourceVersioner
+	policyLastSyncResourceVersioner LastSyncResourceVersioner
+	skip                            skipSynchronizer
+	lastState                       string
+
+	userSubjectRecordStore  cache.Store
+	groupSubjectRecordStore cache.Store
+
+	syncResources func() (sets.String, error)
+
+	group    string
+	resource string
+
+	watchers    []CacheWatcher
+	watcherLock sync.Mutex
+}
+
+func NewAuthCache(clusterRoleInformer rbacv1informers.ClusterRoleInformer,
+	clusterRolebindingInformer rbacv1informers.ClusterRoleBindingInformer,
+	group, resource string,
+	lastSyncResourceVersioner LastSyncResourceVersioner,
+	syncResourcesFunc func() (sets.String, error),
+) *AuthCache {
+	scrLister := syncedClusterRoleLister{
+		clusterRoleInformer.Lister(),
+		clusterRoleInformer.Informer(),
+	}
+	scrbLister := syncedClusterRoleBindingLister{
+		clusterRolebindingInformer.Lister(),
+		clusterRolebindingInformer.Informer(),
+	}
+	result := &AuthCache{
+		clusterRoleLister:               scrLister,
+		clusterRolebindingLister:        scrbLister,
+		syncResources:                   syncResourcesFunc,
+		lastSyncResourceVersioner:       lastSyncResourceVersioner,
+		policyLastSyncResourceVersioner: unionLastSyncResourceVersioner{scrLister, scrbLister},
+
+		group:    group,
+		resource: resource,
+
+		userSubjectRecordStore:  cache.NewStore(subjectRecordKeyFn),
+		groupSubjectRecordStore: cache.NewStore(subjectRecordKeyFn),
+
+		skip: &statelessSkipSynchronizer{},
+
+		watchers: []CacheWatcher{},
+	}
+
+	return result
+}
+
+// synchronize runs a a full synchronization over the cache data.  it must be run in a single-writer model, it's not thread-safe by design.
+func (ac *AuthCache) synchronize() {
+	// if none of our internal reflectors changed, then we can skip reviewing the cache
+	skip, currentState := ac.skip.SkipSynchronize(ac.lastState, ac.lastSyncResourceVersioner, ac.policyLastSyncResourceVersioner)
+	if skip {
+		return
+	}
+
+	userSubjectRecordStore := ac.userSubjectRecordStore
+	groupSubjectRecordStore := ac.groupSubjectRecordStore
+
+	resources, err := ac.syncResources()
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	ac.allResources = resources
+	ac.synchronizeClusterRoleBindings(userSubjectRecordStore, groupSubjectRecordStore)
+	ac.lastState = currentState
+	klog.V(2).Infoln("synchronize...", ac.allResources, ac.allUsers, ac.allGroups)
+}
+
+// synchronizeRoleBindings synchronizes access over each role binding
+func (ac *AuthCache) synchronizeClusterRoleBindings(userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store) {
```

Comment:
More doc here please
1. how to handle role create/update
2. how to handle rolebinding create/update
3. how to handle role/rolebinding delete

Reply:
done